### ISSUE TITLE
udev kmod builtin uses out of date module context

### ISSUE THREAD

**Issue by ddstreet** [systemd/systemd#6723]
udev's rules use a built-in 'kmod' instead of the system modprobe/insmod, and this built-in kmod only validates/refreshes its kmod 'context' every 3 seconds (or longer) during event processing.

However, because other parts of the system rely on udev to load modules correctly, it is not acceptable for it to use an out of date module context. For example, during a system installation:

-the system boots with kernel and initrd with a reduced set of modules, not including nvme module
-udevd starts, and creates its kmod module context, which does not include nvme module
-system installer adds package containg nvme module
-system installer immediately calls udevadm trigger
-udevd sees its kmod module context is not more than 3 seconds old, and does not update it
-udevd finds NVMe pci modalias and asks kmod builtin to load matching driver
-udevd kmod builtin does not find NVMe pci modalias because its context is out of date

this results in the system installer complaining to the user that it found no disks, even though there is a NVMe drive in the system, and the nvme module is installed in the system.

This is not always easily reproducable during system installation or at other times, because udev also reloads all its builtins when any of them are out of date, or when its own systemd/udev conf files are updated.  However on certain systems it is reproducable every time when attempting to install.  Also, it is easily reproduced using this script:

#!/bin/bash
MOD_DIR=/lib/modules/$( uname -r )/kernel/drivers/nvme/host
modprobe -rq nvme
mv $MOD_DIR/nvme.ko .
depmod -a
sleep 3
udevadm trigger
sleep 1
mv nvme.ko $MOD_DIR/
depmod -a
udevadm trigger
sleep 3
grep -q nvme /proc/partitions && echo PASS || echo FAIL

that script does:
1) remove nvme module from the system, reproducing situation where nvme module had not yet been installed
2) waits 3 seconds, because the udev kmod validation timeout is 3 seconds
3) triggers udev, which forces it to reload its kmod context (this could be done with udevadm control -R instead)
4) waits 1 second for the udev trigger to finish, then puts the nvme module back into the system, reproducing the initial installation of the deb/udeb containing the nvme module
5) immediately triggers udev, which should load the nvme module when it sees the nvme pci device
6) wait 3 seconds for udev trigger to finish (plenty of time), and check if the nvme module was loaded

this script reproduces the error every time, due to the stale kmod context. With a fixed udev, this should succeed in loading the nvme module.

I am testing a proposed patch to fix this and I'll provide a merge request soon.

**Comment by ddstreet** [comment]
note that the test script in the description must be run as root on a system that does have a nvme drive that isn't in use (because the nvme module can't be removed if any nvme drive is in use).

**Comment by ddstreet** [comment]
slight update to test script, add a check to make sure the nvme module was unloaded - if any nvme partitions are mounted it would cause a false positive result.

#!/bin/bash
MOD_DIR=/lib/modules/$( uname -r )/kernel/drivers/nvme/host
modprobe -rq nvme
mv $MOD_DIR/nvme.ko .
depmod -a
sleep 3
udevadm trigger
sleep 1
mv nvme.ko $MOD_DIR/
depmod -a
grep -q nvme /proc/partitions && echo FAIL nvme driver still loaded unmount all nvme partitions && exit 1 
udevadm trigger
sleep 3
grep -q nvme /proc/partitions && echo PASS || echo FAIL

**Comment by poettering** [comment]
not sure i get what the point of all of this. but isn't "udevadm control --reload" after dropping in your stuff enough?

**Comment by ddstreet** [comment]
> not sure i get what the point of all of this. but isn't "udevadm control --reload" after dropping
> in your stuff enough?

it isn't someone else's job to tell udev that the installed modules have changed.  And the installed modules aren't static, especially during system installation when the installer adds modules and then expects udev to correctly load those newly installed modules.  See my explanation in the description of how udev breaks system installation for some systems installing on nvme drives.  The point of this is to ensure udev always uses the correct builtin backend data.

if you think the builtin backends are static, why do you have them checked every 3 seconds (during udev activity)?  is checking their validity for each builtin call going to crush performance?  it's a simple check on file stat time info.

**Comment by poettering** [comment]
if you think the builtin backends are static, why do you have them checked every 3 seconds (during udev activity)? is checking their validity for each builtin call going to crush performance? it's a simple check on file stat time info.

Well, "udevadm control --reload" is the way to go for this, and that's what RPMs for example generally issue to get their hwdb/kmod bits noticed by udev. I am not a fan of automatic configuration reload I must say, due to the inherent raciness. I'd rather drop the rechecking altogether than to remove the 3s thing.

Please, use the offical API for reloading udev if you want your stuff noticed.

**Comment by ddstreet** [comment]
> Well, "udevadm control --reload" is the way to go for this, and that's what RPMs for example generally
> issue to get their hwdb/kmod bits noticed by udev. I am not a fan of automatic configuration reload I
> must say, due to the inherent raciness. I'd rather drop the rechecking altogether than to remove the 3s
> thing.

Ok, I'll send a patch to remove udev's builtin rechecking entirely, and we'll see how many bug reports start getting opened xD

> Please, use the offical API for reloading udev if you want your stuff noticed.

I'm not sure why you think that everyone else in the world that isn't systemd, needs to understand systemd internals (like that it keeps track of installed modules but only checks if it's valid every 3 seconds).  Sure, I can change the debian installer to work around *this specific case* where udev is broken by telling udev its kmod context is out of date (by forcing udev to kill and recreate all its manager forked processes, and reload its own udev config data, which is way, way more heavyweight than simply having each builtin check its own correctness and reload only itself).  However, that means this bug will continue to exist in udev, and other people will continue to keep hitting it, wasting waaaaay too much of their time debugging it, and then either they'll try to fix it (which I assume you will reject) and/or they'll also work around the problem for the very specific case it's failing for them.

The bottom line here is: expecting other people to know they need to call udevadm -R is a broken assumption.  It is made even worse by the 3-second automatic recheck/reload, which hides the fact that udev expects *other* code to call udevadm -R when anything changes.  What is really means is "you must call udevadm -R when anything udev uses changes, or you must wait 3 seconds".

This is simply a bug waiting to continue happening, and really I suggest either udev's 3-second recheck/reload is removed completely - requiring other code to *always* call udevadm -R, so this bug won't be so hidden - or udev should manage its rechecks/reloads correctly, without a 3 second window where udev does not work right.

**Comment by poettering** [comment]
I talked to Kay about all of this, and he agrees too that this really should be handles with explicit reloads rather than automatic implicit ones. This has been done like this since a long time and everything more magic tends to be racy.

Sorry for not responding more timely to this. Closing.

**Comment by ddstreet** [comment]
> I talked to Kay about all of this, and he agrees too that this really should be handles with explicit
> reloads rather than automatic implicit ones. 

Uh, you do realize that udevd CURRENTLY does automatic implicit reloads right?  It just does it with a 3-second window of stale data, instead of doing implicit automatic reloads before any stale data is used.

> This has been done like this since a long time and
> everything more magic tends to be racy.

Racy?  You realize all the builtin code is completely thread unsafe, right?  But that's ok because it's only ever called from fork()ed single-threaded worker threads, so there is no possiblity of races?

Do whatever you want, but your excuse doesn't make much sense.  I certainly don't know what you mean by "magic", since this patch is anything but magic.  Maybe Kay can comment more specifically in this issue?

**Comment by xnox** [comment]
@poettering This is an interesting issue, and I have learned something new from it. Looking at `src/core/macros.systemd.in` it looks like `%udev_hwdb_update` only updates the hwdb, but doesn't reload udevadm shouldn't `%udev_hwdb_update` also call `%udev_rules_update`? Are there any recommendations or suggestions as to when `udevadm control --reload` should be called? It seems like possibly in Debian/Ubuntu dpkg trigger should be installed to call `udevadm control --reload` whenever new udev rules are installed, or additional kernel modules are installed? (e.g. dkms compiled modules, or modules that are split into separate .deb / .udeb)
